### DISCUSSION TITLE
HELP with this error: stack expects each tensor to be equal size, but got [1, 32, 32, 16] at entry 0 and [1, 32, 32, 15] at entry 7

### DISCUSSION THREAD

**Discussion by vlgonz** [Project-MONAI/tutorials#1675] [General]
Hi, I have **the title error and I do not know how to solve it**. This is MY code: 
keys=("image", "label")

xforms = [
        LoadImaged(keys, ensure_channel_first=True, image_only=True),
        Orientationd(keys, axcodes="LPS"),
        CropForegroundd(keys, source_key=keys[0]),
        Spacingd(keys, pixdim=(1.0, 1.0, 3.0), mode=("bilinear", "nearest")[: len(keys)]),
        ScaleIntensityRanged(keys[0], a_min=-1000.0, a_max=500.0, b_min=0.0, b_max=1.0, clip=True),
        **SpatialPadd(keys, spatial_size=(32, 32, 16), mode="constant"),**
    ]
    if mode == "train":
        xforms.extend(
            [
                **RandCropByPosNegLabeld(keys, label_key=keys[1], spatial_size=(32, 32, 16), num_samples=2)**,
                RandGaussianNoised(keys[0], prob=0.15, std=0.01),
                RandFlipd(keys, spatial_axis=0, prob=0.5),
                RandFlipd(keys, spatial_axis=1, prob=0.5),
                RandFlipd(keys, spatial_axis=2, prob=0.5),
            ]
Why do I get that error if I have added SpatialPadd with a spatial size of (32,32,16) before RandCropByPosNegLabeld? Thank you

**Comment by ericspod** [comment]
I'm not sure I can duplicate the issue here. What size are your tensors just before `SpatialPadd`? In your `CropForegroundd` definition I think you want `source_key=keys[1]` instead? I have the following which works, see if there's anything you can change to reproduce your problem:

```python
import monai, torch
import monai.transforms as mt

img = torch.rand(1, 32, 32, 17)
lbl = torch.zeros_like(img)
lbl[0, 10:20, 10:20, 4:12] = 1  # one small label
data = {"image": img, "label": lbl}

keys = ("image", "label")
com = mt.Compose(
    [
        mt.CropForegroundd(keys, source_key=keys[1]),
        mt.SpatialPadd(keys, spatial_size=(32, 32, 16), mode="constant"),
        mt.RandCropByPosNegLabeld(keys, label_key=keys[1], spatial_size=(32, 32, 16), num_samples=2),
        mt.RandGaussianNoised(keys[0], prob=0.15, std=0.01),
        mt.RandFlipd(keys, spatial_axis=0, prob=0.5),
        mt.RandFlipd(keys, spatial_axis=1, prob=0.5),
        mt.RandFlipd(keys, spatial_axis=2, prob=0.5),
    ]
)
ds = monai.data.Dataset([data], com)
dl = monai.data.DataLoader(ds)
res = monai.utils.first(dl)
print(res["image"].shape, res["label"].shape)  # prints torch.Size([2, 1, 32, 32, 16]) torch.Size([2, 1, 32, 32, 16])

```

**Comment by vlopezgz** [comment]
Hi, thanks for your response. Originally, each tensor had dimensions of (512, 512, z), but after applying CropForeground, each tensor has a different size due to variations in the background of each image. I opted not to use CropForeground with source_key=keys[1] because it creates a patch of the region I intend to segment, which is not my objective. Instead, I aim to eliminate the black areas surrounding the main body, ensuring that the input data to the network does not include a large black portion. To achieve this, I removed CropForeground and applied CenterSpatialCrop with a fixed size using roi_size=(460, 460, -1). This adjustment resolved the error, although it did not eliminate the black parts from all images.

**Comment by ericspod** [comment]
I'm still not sure how this error is coming up for you. `CropForegroundd` will remove the black areas around your image which will possibly be resized by `Spacingd`, but then `SpatialPadd` should ensure that the image is no smaller than (32,32,16). This is what I get in my example if you play with the input sizes. `RandCropByPosNegLabeld` should also then get an input that's at least that size and be able to find samples of the right size. I can only suggest using my example code with different combinations of transforms to replicate your setup and with your data to see what reproduces the problem.

**Reply by AyaAl1ah** [reply]
Hi I think I have a similar issue. I changed a few things like spatial_size=(30, 30, 30) and img_size=(30, 30, 30) in the model as well as in the evaluation function. Also I have two classes and changed it out_channels=2,
    img_size=(30, 30, 30), and also the feature size 
    feature_size=10
the out put:
Training (X / X Steps) (loss=X.X):   0%|          | 0/20 [00:00<?, ?it/s]

Image size: torch.Size([4, 1, 30, 30, 30]) # i printed the shapes in the training loop
Label size: torch.Size([4, 1, 30, 30, 30])

Training (X / X Steps) (loss=X.X):   0%|          | 0/20 [00:06<?, ?it/s]
RuntimeError                              Traceback (most recent call last)
[<ipython-input-14-60cc62a0eff2>](https://localhost:8080/#) in <cell line: 70>()
     69 metric_values = []
     70 while global_step < max_iterations:
---> 71     global_step, dice_val_best, global_step_best = train(global_step, train_loader, dice_val_best, global_step_best)
     72 model.load_state_dict(torch.load(os.path.join(root_dir, "best_metric_model.pth")))

8 frames
[/usr/local/lib/python3.10/dist-packages/torch/_tensor.py](https://localhost:8080/#) in __torch_function__(cls, func, types, args, kwargs)
   1441 
   1442         with _C.DisableTorchFunctionSubclass():
-> 1443             ret = func(*args, **kwargs)
   1444             if func in get_default_nowrap_functions():
   1445                 return ret

RuntimeError: Sizes of tensors must match except in dimension 1. Expected size 16 but got size 30 for tensor number 1 in the list. I get this error and nothing seems to fix it. sorry if I am bothering you but I am still learning thanks.
